libr/include/r_cons.h

```diff
@@ -944,6 +950,21 @@ R_API int r_line_set_hist_callback(RLine *line, RLineHistoryUpCb cb_up, RLineHis
 R_API int cmd_history_up(RLine *line);
 R_API int cmd_history_down(RLine *line);
 
+R_API void r_line_completion_init(RLineCompletion *completion);
+R_API void r_line_completion_fini(RLineCompletion *completion);
+R_API void r_line_completion_push_owned(RLineCompletion *completion, char *str);
+R_API void r_line_completion_push_weak(RLineCompletion *completion, const char *str);
```

Comment:
i would simplify and just have a push(), if you mix owned and weaks in the same pvector it will end up leaking

Reply:
They don't leak as long as the correct functions are used. The way it works is this: An empty RLineCompletion always has its `args_weak == true`. As long as you push weak refs into it, this will stay. As soon as you push an owned string, all currently saved strings will be copied and `args_weak` set to true. When you then push a weak ref, it will be copied too. On clearing, depending on `args_weak`, either all strings are freed or left alone.

Reply:
Mmm why do we need owned/weak? Can't we either have all strings weak or all strings owned? It seems a bit messy...

Reply:
See the last comment. Only weak can't work and only owned would need a lot of copying.